From f32e520cf12956f98f19a0a1a8b40d8d761cb19b Mon Sep 17 00:00:00 2001
From: Harini Katakam <harini.katakam@amd.com>
Date: Mon, 27 Nov 2023 13:34:54 +0530
Subject: [LINUX PATCH] net: axienet: Add support for 2.5 using phylink
 framework

Support use of 2500BASEX mode. Dont warn if switch_x_sgmii is used
without 1GBX or SGMII phy-mode.

Signed-off-by: Harini Katakam <harini.katakam@amd.com>
---
 .../net/ethernet/xilinx/xilinx_axienet_main.c    | 16 +++++++++++-----
 1 file changed, 11 insertions(+), 5 deletions(-)

diff --git a/drivers/net/ethernet/xilinx/xilinx_axienet_main.c b/drivers/net/ethernet/xilinx/xilinx_axienet_main.c
index 256c4f050bce..1ee583888b8d 100644
--- a/drivers/net/ethernet/xilinx/xilinx_axienet_main.c
+++ b/drivers/net/ethernet/xilinx/xilinx_axienet_main.c
@@ -3011,7 +3011,8 @@ static struct phylink_pcs *axienet_mac_select_pcs(struct phylink_config *config,
 	struct axienet_local *lp = netdev_priv(ndev);
 
 	if (interface == PHY_INTERFACE_MODE_1000BASEX ||
-	    interface ==  PHY_INTERFACE_MODE_SGMII)
+	    interface ==  PHY_INTERFACE_MODE_SGMII ||
+	    interface == PHY_INTERFACE_MODE_2500BASEX)
 		return &lp->pcs;
 
 	return NULL;
@@ -3602,13 +3603,13 @@ static int axienet_probe(struct platform_device *pdev)
 				goto cleanup_clk;
 		}
 	}
-	if (lp->switch_x_sgmii && lp->phy_mode != PHY_INTERFACE_MODE_SGMII &&
+/*	if (lp->switch_x_sgmii && lp->phy_mode != PHY_INTERFACE_MODE_SGMII &&
 	    lp->phy_mode != PHY_INTERFACE_MODE_1000BASEX) {
 		dev_err(&pdev->dev, "xlnx,switch-x-sgmii only supported with SGMII or 1000BaseX\n");
 		ret = -EINVAL;
 		goto cleanup_clk;
 	}
-
+*/
 	/* Set default USXGMII rate */
 	lp->usxgmii_rate = SPEED_1000;
 	of_property_read_u32(pdev->dev.of_node, "xlnx,usxgmii-rate",
@@ -3860,8 +3861,13 @@ static int axienet_probe(struct platform_device *pdev)
 
 	__set_bit(lp->phy_mode, lp->phylink_config.supported_interfaces);
 	if (lp->switch_x_sgmii) {
-		__set_bit(PHY_INTERFACE_MODE_1000BASEX,
-			  lp->phylink_config.supported_interfaces);
+		if (lp->phy_mode == PHY_INTERFACE_MODE_2500BASEX) {
+			__set_bit(PHY_INTERFACE_MODE_2500BASEX,
+				  lp->phylink_config.supported_interfaces);
+		} else {
+			__set_bit(PHY_INTERFACE_MODE_1000BASEX,
+				  lp->phylink_config.supported_interfaces);
+		}
 		__set_bit(PHY_INTERFACE_MODE_SGMII,
 			  lp->phylink_config.supported_interfaces);
 	}
-- 
2.17.1

